### PR TITLE
Add response_format support for openai and a way to set the functions

### PR DESCRIPTION
**Pull request by Arttii**
## Description

I needed support for the JSON output that OpenAI supports by setting the type. This is already supported by the underlying client.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] No tests

## Checklist:

- [X] I have read the [Contributing documentation](https://github.com/henomis/lingoose/blob/main/CONTRIBUTING.md).
- [ X] I have read the [Code of conduct documentation](https://github.com/henomis/lingoose/blob/main/CODE_OF_CONDUCT.md).
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by BabySid** [comment]
hi, I do need this feature too. Is there a plan to merge this PR? thanks @henomis

**Comment by henomis** [comment]
> hi, I do need this feature too. Is there a plan to merge this PR? thanks @henomis

@BabySid this PR needs some refactor.

**Comment by BabySid** [comment]
> > hi, I do need this feature too. Is there a plan to merge this PR? thanks @henomis
> 
> @BabySid this PR needs some refactor.

got. Thanks for the reply

**Comment by Arttii** [comment]
@henomis please let me know, what you think should be changed. Would be happy to incorporate it.

**Comment by henomis** [comment]
> @henomis please let me know, what you think should be changed. Would be happy to incorporate it.

@Arttii just follow the review comments.

File: go.sum [hunk 1]
```diff
@@ -27,6 +27,8 @@ github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZb
 github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
 github.com/sashabaranov/go-openai v1.19.2 h1:+dkuCADSnwXV02YVJkdphY8XD9AyHLUWwk6V7LB6EL8=
 github.com/sashabaranov/go-openai v1.19.2/go.mod h1:lj5b/K+zjTSFxVLijLSTDZuP7adOgerWeFyZLUhAKRg=
+github.com/sashabaranov/go-openai v1.21.0 h1:isAf3zPSD3VLd0pC2/2Q6ZyRK7jzPAaz+X3rjsviaYQ=
+github.com/sashabaranov/go-openai v1.21.0/go.mod h1:lj5b/K+zjTSFxVLijLSTDZuP7adOgerWeFyZLUhAKRg=
```
Comment (henomis) [inline go.sum:31]: This is out of scope of this PR.

File: llm/openai/common.go [hunk 2]
```diff
@@ -33,6 +33,8 @@ const (
 	GPT4TurboPreview      Model = openai.GPT4TurboPreview
 	GPT4VisionPreview     Model = openai.GPT4VisionPreview
 	GPT4                  Model = openai.GPT4
+	GPT4Turbo             Model = openai.GPT4Turbo
+	GPT4Turbo20240409     Model = openai.GPT4Turbo20240409
```
Comment (henomis) [inline llm/openai/common.go:37]: Out of Scope

File: llm/openai/openai.go [hunk 3]
```diff
@@ -324,7 +324,7 @@ func (o *OpenAI) getChatCompletionRequestTools() []openai.Tool {
 	for _, function := range o.functions {
 		tools = append(tools, openai.Tool{
 			Type: "function",
-			Function: openai.FunctionDefinition{
+			Function: &openai.FunctionDefinition{
```
Comment (henomis) [inline llm/openai/openai.go:327]: Out of scope

**Review by henomis — CHANGES_REQUESTED** [review]

File: llm/openai/openai.go [hunk 4]
```diff
@@ -37,6 +37,7 @@ type OpenAI struct {
 	streamCallbackFn StreamCallback
 	toolChoice       *string
 	cache            *cache.Cache
+	responseFormat   openai.ChatCompletionResponseFormatType
```
[the remark this answers is not in the document]
Comment (Arttii) [reply llm/openai/openai.go:40]: Should we use a local alias for this or what is your preference?

File: thread/thread.go [hunk 5]
```diff
@@ -208,7 +208,7 @@ func (c *Content) Format(input types.M) *Content {
 	prompt := prompt.NewPromptTemplate(c.Data.(string))
 	err := prompt.Format(input)
 	if err != nil {
-		return c
+		panic(c)
```
[the remark this answers is not in the document]
Comment (Arttii) [reply thread/thread.go:212]: This is  an interesting question as we should at least surface the error. This tripped me up a few times as the llm was just being prompted with an empty prompt essentially. Should we log or what is your preference?

File: go.sum (hunk 1 above)
Comment (Arttii) [reply go.sum:31]: Fair enough, I will remove this.

File: llm/openai/openai.go [hunk 6]
```diff
@@ -75,11 +76,22 @@ func (o *OpenAI) WithClient(client *openai.Client) *OpenAI {
 	return o
 }
 
+// WithFunctions sets the functions to use for the OpenAI instance.
+func (o *OpenAI) WithFunctions(functions map[string]Function) *OpenAI {
+	o.functions = functions
+	return o
+}
+
```
[the remark this answers is not in the document]
Comment (Arttii) [reply llm/openai/openai.go:84]: Why is this out of scope?

File: llm/openai/openai.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (henomis) [reply llm/openai/openai.go:40]: Yes use an alias, see the comment below.

File: llm/openai/openai.go [hunk 7]
```diff
@@ -75,11 +76,22 @@ func (o *OpenAI) WithClient(client *openai.Client) *OpenAI {
...
 func (o *OpenAI) WithToolChoice(toolChoice *string) *OpenAI {
 	o.toolChoice = toolChoice
 	return o
 }
 
+func (o *OpenAI) WithResponseFormat(format openai.ChatCompletionResponseFormatType) *OpenAI {
```
[the remark this answers is not in the document]
Comment (henomis) [reply llm/openai/openai.go:90]: Lingoose's users must deal with lingoose types.

File: llm/openai/openai.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (henomis) [reply llm/openai/openai.go:84]: I'm not 100% sure of this method and this change must be discussed.

**Review by henomis — CHANGES_REQUESTED** [review]

**Comment by BabySid** [comment]
Hi, @henomis  It seems that this PR has been on hold for a long time. Does the lingoose project have any plans to develop this feature? To be honest, this feature is still very important for openai.  I also know that this PR is currently in a very awkward situation, and the author may not be able to continue development due to some unknown reasons. It would be a very commendable thing if the project can integrate this feature.

**Comment by henomis** [comment]
@BabySid 
I'll open a new PR for this feature.

**Comment by henomis** [comment]
Implemented here: https://github.com/henomis/lingoose/pull/199

**Pull request closed without merging**